### PR TITLE
Task #39 Code New Record Input Screen and Input Validation

### PR DESCRIPTION
**Pull request by domlau**
@taerakwon @kkh0811 @mekhaleganesh  @RyanYizhaoHuang @prtk15 

All validation have been tested for user input and completed. Record is saved to DB when validation passes.

### CODE REVIEW

File: PasswordApplication/AddUserRecordhelper.cs [hunk 1]
```diff
@@ -45,9 +43,9 @@ public bool SaveEntity(AbDatabaseEntity Entity)
             }
             finally
             {
-                SQLconn.Connect().Close();
+                conn.Close();
             }
-            throw new NotImplementedException();
+            return true;
```
Comment (RyanYizhaoHuang) [inline PasswordApplication/AddUserRecordhelper.cs]: it is better to check the **addUserRecord function(SaveEntity)** true or not by examining the ExecuteeNonQuery.() return how many records are impacted. 
If you return true at then end, that means if there are no exception the method will always return true. By using checking the return value of the **int ExecuteNonQuery() **, you know how many records are save to the DB.
If you save record to DB but no row is impacted, something wrong. 
Not a big thing of you code. Just remind you the other way I did to check the function success or not.

File: PasswordApplication/AddUserRecordhelper.cs [hunk 2]
```diff
@@ -13,26 +13,24 @@ namespace PasswordApplication
...
                 //calling methods to demonstrate sqlCommand capabilities
-                addUserRecord = new SqlCommand("dbo.AddRecord", SQLconn.Connect());
+                addUserRecord = new SqlCommand("dbo.AddRecord", conn);
                 addUserRecord.CommandType = CommandType.StoredProcedure;
-                
-                //addUserRecord.Parameters.Add("@categoryName", SqlDbType.VarChar).Value = 
+                addUserRecord.Parameters.Add("@categoryID", SqlDbType.VarChar).Value = category.CategoryID;
                 addUserRecord.Parameters.Add("@id", SqlDbType.VarChar).Value = userRecord.UserName;
                 addUserRecord.Parameters.Add("@pw", SqlDbType.VarChar).Value = userRecord.UserPassword;
                 addUserRecord.Parameters.Add("@note", SqlDbType.VarChar).Value = userRecord.Note;
```
Comment (RyanYizhaoHuang) [inline PasswordApplication/AddUserRecordhelper.cs]: does it save the UserAccountId to the UserRecord table? if we have more than more UserAccount. We need to keep tracking which **User** is saving the record. And the UserRecord table has a field is UserAccountID.

File: PasswordApplication/UserNameValidator.cs [hunk 3]
```diff
@@ -16,25 +16,36 @@ namespace PasswordApplication
     /// </summary>
     class UserNameValidator : IValidation
     {
-        // create pattern
-        private string pattern = " ^[a - z | A - Z |] +[a - z | A - Z | 0 - 9] * ";
-        //Regular expression
+        // create pattern to match username
+        //regular expression only allows user to input alphabets or numbers only.
+        private string pattern = "^[a-z|A-Z|]+[a-z|A-Z|0-9]*";
```
Comment (RyanYizhaoHuang) [inline PasswordApplication/UserNameValidator.cs]: The function is ok, just leave a reminder here, the pattern should accept Email address at least. Our first user will be Centennial people, one of the user names for them is my centennial email address. We can change later. Just remember.

File: PasswordApplication/AddUserRecordhelper.cs (hunk 1 above)
Comment (domlau) [inline PasswordApplication/AddUserRecordhelper.cs]: @RyanYizhaoHuang 
That makes sense. Thank you for pointing that out. I will make the changes. 👍

File: PasswordApplication/UserNameValidator.cs (hunk 3 above)
Comment (domlau) [inline PasswordApplication/UserNameValidator.cs]: Oh right, I will make change to the validation string pattern to accommodate E-mail addresses as well.

File: PasswordApplication/AddUserRecordhelper.cs (hunk 2 above)
Comment (domlau) [inline PasswordApplication/AddUserRecordhelper.cs]: @RyanYizhaoHuang I will have to talk to @kkh0811 regarding to this issue.

**Comment by domlau** [comment]
@RyanYizhaoHuang @kkh0811 @mekhaleganesh @prtk15 I have made some adjustments to the codes. Please refer to commit bd0bf1ab0de1a17ff127a5228213baedc12fe8b5

**Comment by domlau** [comment]
Seems like there is a bit of an issue with the DB part when creating a new user record(Duplicates record when returned to main form). Will have to fix this issue first. @taerakwon Should I close this pull request now and make my changes first and then reopen the pull request?

File: PasswordApplication/NewRecordForm.cs [hunk 4]
```diff
@@ -54,27 +53,33 @@ private void NewRecordForm_Load(object sender, EventArgs e)
...
+            //if statement of the return bool value from method validating
             if (validating() == true)
             {
                 //pass information to DB
+                //instantiate AddUserRecordHelper, UserRecord, Category
                 AddUserRecordhelper addUserRecordhelper = new AddUserRecordhelper();
                 UserRecord addRecord = new UserRecord();
                 Category category = new Category();
-                //UserRecordExtend addUserRecordExtend = new UserRecordExtend();
                 category.CategoryID = CategoryOptionComboBox.SelectedIndex;
```
Comment (RyanYizhaoHuang) [inline PasswordApplication/NewRecordForm.cs]: @domlau 
Just find one issue if you use CategoryOptionComboBox.SelectedIndex as the paratemter for CategoryId. This is not the exactly CategoryId from the DB, it is the ID when you populate the ComboBox. When you have multiple categoryId but their ID in database is not consecutive (like our DB now), you won't get the correct categoryId.
Try to save the ID first, my example for treeView:
TreeNode NewNode = new TreeNode(row["CategoryName"].
//Save the CategoryID to the treeNode 
NewNode.Tag = row["CategoryID"];
CategoryTreeView.Nodes[0].Nodes.Add(NewNode);

**Pull request closed without merging**